src/GraphQL/ConstrainedArray.cs

```diff
@@ -0,0 +1,68 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace GraphQL
+{
+    // Represents an array based on another array of equal or greater length.
+    // The wrapped array is usually obtained from ArrayPool.
+    internal sealed class ConstrainedArray<T> : IList<T>, IList
+    {
+        private readonly T[] _array;
+
+        public ConstrainedArray(T[] array, int count)
+        {
+            if (count < 0 || count > array.Length)
+                throw new ArgumentOutOfRangeException(nameof(count));
+
+            _array = array;
+            Count = count;
+        }
+
+        public T this[int index] { get => _array[index]; set => throw new NotSupportedException(); }
+
+        object IList.this[int index] { get => _array[index]; set => throw new NotImplementedException(); }
+
+        public int Count { get; }
+
+        public bool IsSynchronized => false;
+
+        public object SyncRoot => _array.SyncRoot;
+
+        public bool IsReadOnly => true;
+
+        public bool IsFixedSize => true;
+
+        public int IndexOf(T item) => throw new NotSupportedException();
+
+        public int IndexOf(object value) => throw new NotSupportedException();
+
+        public bool Contains(T item) => throw new NotSupportedException();
```

Comment:
This and other methods that do not change the state of the array can be implemented. Until I just didn’t do it, because it’s not necessary.